File: MatriX/samples/csharp/MiniClient/FrmSendFile.cs
using System.IO;
using System.Windows.Forms;
using Matrix;
using Matrix.Xmpp.Client;

namespace MiniClient
{
    public partial class FrmSendFile : Form
    {
        private string sid = "";
        private Jid _jid;
        private FileTransferManager fm;
        public FrmSendFile(FileTransferManager ftm, Jid jid)
        {
            InitializeComponent();

            fm = ftm;
            _jid = jid;

            Text = "File transfer: " + jid;

            fm.OnError += fm_OnError;
            fm.OnEnd += fm_OnEnd;
            fm.OnStart += fm_OnStart;
            fm.OnProgress += fm_OnProgress;
        }

        void fm_OnError(object sender, ExceptionEventArgs e)
        {
            var ex = e.Exception as FileTransferException;
            if (ex.Sid != sid)
                return;

            // file transfer failed because our contact went offline or some
            // other errror occured.
            MessageBox.Show("file transfer failed!", "Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
            cmdChooseFile.Enabled = cmdAbort.Enabled = cmdSend.Enabled = false;
        }

        void fm_OnEnd(object sender, FileTransferEventArgs e)
        {
            if (e.Sid != sid)
                return;

            MessageBox.Show("file transfer ended with success!", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information);
            Close();
        }

        void fm_OnStart(object sender, FileTransferEventArgs e)
        {
            if (e.Sid != sid)
                return;
            /// file transfer started
        }

        void fm_OnProgress(object sender, FileTransferEventArgs e)
        {
            if (e.Sid != sid)
                return;

            progressBar.Value = (int) (((double)e.BytesTransmitted / (double)e.FileSize) * 100);
        }

        private void cmdChooseFile_Click(object sender, System.EventArgs e)
        {
            var of = new OpenFileDialog();
            if (of.ShowDialog() == DialogResult.OK)
            {
                lblFileName.Text = of.FileName;
                
                var fi = new FileInfo(of.FileName);
                lblSize.Text = Util.HumanReadableFileSize(fi.Length);

                cmdSend.Enabled = true;
            }
        }

        
        private void cmdSend_Click(object sender, System.EventArgs e)
        {
            sid = fm.Send(_jid, lblFileName.Text, txtDescription.Text);
            cmdSend.Enabled = cmdChooseFile.Enabled = false;
        }
    }
}

File: MatriX/samples/vb/MiniClient/FrmSendFile.vb
Imports System.IO
Imports System.Windows.Forms
Imports Matrix
Imports Matrix.Xmpp.Client

Public Partial Class FrmSendFile
    Inherits Form
    Private sid As String = ""
    Private _jid As Jid
    Private WithEvents fm As FileTransferManager
		
    Public Sub New(ftm As FileTransferManager, jid As Jid)
        InitializeComponent()

        fm = ftm
        _jid = jid

        Text = "File transfer: " + jid.ToString

        AddHandler fm.OnError, AddressOf fm_OnError
        AddHandler fm.OnEnd, AddressOf fm_OnEnd
        AddHandler fm.OnStart, AddressOf fm_OnStart
        AddHandler fm.OnProgress, AddressOf fm_OnProgress
    End Sub

    Private Sub fm_OnError(sender As Object, e As ExceptionEventArgs)
        Dim ex = TryCast(e.Exception, FileTransferException)
        If ex.Sid <> sid Then
            Return
        End If

        ' file transfer failed because our contact went offline or some
        ' other errror occured.
        MessageBox.Show("file transfer failed!", "Failure", MessageBoxButtons.OK, MessageBoxIcon.[Error])
        cmdChooseFile.Enabled = InlineAssignHelper(cmdAbort.Enabled, InlineAssignHelper(cmdSend.Enabled, False))
    End Sub

    Private Sub fm_OnEnd(sender As Object, e As FileTransferEventArgs)
        If e.Sid <> sid Then
            Return
        End If

        MessageBox.Show("file transfer ended with success!", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Close()
    End Sub

    Private Sub fm_OnStart(sender As Object, e As FileTransferEventArgs)
        If e.Sid <> sid Then
            Return
        End If
        ''' file transfer started
    End Sub

    Private Sub fm_OnProgress(sender As Object, e As FileTransferEventArgs)
        If e.Sid <> sid Then
            Return
        End If

        progressBar.Value = CInt((CDbl(e.BytesTransmitted) / CDbl(e.FileSize)) * 100)
    End Sub

    Private Sub cmdChooseFile_Click(sender As Object, e As System.EventArgs) Handles cmdChooseFile.Click 
        Dim ofd = New OpenFileDialog()
        If ofd.ShowDialog() = DialogResult.OK Then
            lblFileName.Text = ofd.FileName

            Dim fi = New FileInfo(ofd.FileName)
            lblSize.Text = Util.HumanReadableFileSize(fi.Length)

            cmdSend.Enabled = True
        End If
    End Sub

    Private Sub cmdSend_Click(sender As Object, e As System.EventArgs) Handles cmdSend.Click
        sid = fm.Send(_jid, lblFileName.Text, txtDescription.Text)
        cmdSend.Enabled = cmdChooseFile.Enabled = False
    End Sub

    Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
        target = value
        Return value
    End Function
End Class